web/controllers/api/v1/message_controller.ex

```diff
@@ -6,26 +6,58 @@ defmodule Kastlex.API.V1.MessageController do
 
   plug Kastlex.Plug.EnsurePermissions
 
-  # TODO: partitioner
   def produce(conn, %{"topic" => topic, "partition" => partition} = params) do
     {partition, _} = Integer.parse(partition)
     key = Map.get(params, "key", "")
     {:ok, value, conn} = read_body(conn)
-    case :brod.produce_sync(:kastlex, topic, partition, key, value) do
-      :ok ->
-        send_resp(conn, 204, "")
+    try_produce(conn, topic, partition, key, value)
+  end
+
+  # no partition, use random partitioner
+  def produce(conn, %{"topic" => topic} = params) do
+    key = Map.get(params, "key", "")
+    {:ok, value, conn} = read_body(conn)
+    case :brod_client.get_partitions_count(:kastlex, topic) do
+      {:ok, partitions_cnt} ->
+        partitions = :lists.seq(0, partitions_cnt - 1) |> Enum.shuffle
+        try_produce(conn, topic, partitions, key, value)
       {:error, :UnknownTopicOrPartition} ->
         send_json(conn, 404, %{error: "Unknown topic or partition"})
-      {:error, {:producer_not_found, _topic}} ->
-        send_json(conn, 404, %{error: "Unknown topic"})
-      {:error, {:producer_not_found, _topic, _partition}} ->
-        send_json(conn, 404, %{error: "Unknown partition"})
       {:error, reason} ->
         Logger.error "#{reason}"
         send_json(conn, 503, %{error: "Service unavailable"})
     end
   end
 
+  defp try_produce(conn, topic, partition, key, value) when is_integer(partition) do
+    try_produce(conn, topic, [partition], key, value, nil)
+  end
+  defp try_produce(conn, topic, partitions, key, value) when is_list(partitions) do
+    try_produce(conn, topic, partitions, key, value, nil)
+  end
+
+  defp try_produce(conn, topic, [], _key, _value, error) do
+    case error do
+      :UnknownTopicOrPartition ->
+        send_json(conn, 404, %{error: "Unknown topic or partition"})
+      {:producer_not_found, _topic} ->
+        send_json(conn, 404, %{error: "Unknown topic"})
+      {:producer_not_found, _topic, _partition} ->
+        send_json(conn, 404, %{error: "Unknown partition"})
+      reason ->
+        Logger.error "Failed to produce to #{topic}: #{reason}"
+        send_json(conn, 503, %{error: "Service unavailable"})
+    end
+  end
+  defp try_produce(conn, topic, [p | partitions], key, value, _last_error) do
```

Comment:
when auto_start_producers is enabled, `produce` call will make `brod_client` try to start the producer, before starting the topic sup,`brod_client` check existence of the topic (and partition), return `{error, ?EC_UNKNOWN_TOPIC_OR_PARTITION}` if not found.